src/main/cpp/SwerveModule.cpp

```diff
@@ -0,0 +1,245 @@
+
+#include "SwerveModule.h"
+#include <iostream>
+
+bool SwerveModule::Init(Location loc) {
+    // PID Controller stuff
+    // drive PID gains
+    double drive_kP = RobotParams::GetParam("swerve.drive_pid.kP", 0.0);
+    double drive_kI = RobotParams::GetParam("swerve.drive_pid.kI", 0.0);
+    double drive_kD = RobotParams::GetParam("swerve.drive_pid.kD", 0.0);
+
+    // steer PID gains
+    double steer_kP = RobotParams::GetParam("swerve.steer_pid.kP", 0.0);
+    double steer_kI = RobotParams::GetParam("swerve.steer_pid.kI", 0.0);
+    double steer_kD = RobotParams::GetParam("swerve.steer_pid.kD", 0.0);
+
+    // create our PIDs
+    drive_pid_ = std::make_shared<frc::PIDController>(drive_kP, drive_kI, drive_kD);
+    steer_pid_ = std::make_shared<frc::PIDController>(steer_kP, steer_kI, steer_kD);
+
+    OKC_CHECK(this->steer_pid_ != nullptr);
+    steer_pid_->EnableContinuousInput(0, 360);
+
+    // units and conversions and numbers and stuff
+    L2_GEAR_RATIO_ = RobotParams::GetParam("swerve.l2_gear_ratio", 0.0);
+    WHEEL_DIAMETER_ = RobotParams::GetParam("swerve.wheel_diameter", 0.0);
```

Comment:
Setting these default to zero may result in divide by zero if params fail to load. Consider a different default like 1.